unix_integration/src/resolver.rs

```diff
@@ -733,27 +735,48 @@ where
         &self,
         token: &Option<UserToken>,
         account_id: &Id,
-        cred: &str,
-    ) -> Result<Option<bool>, ()> {
+        cred: Option<PamCred>,
+    ) -> Result<ClientResponse, ()> {
         debug!("Attempt online password check");
+        // Unwrap the cred for passing to the step function
+        let ucred = match &cred {
+            Some(PamCred::Password(v)) => Some(v.clone()),
+            Some(PamCred::MFACode(v)) => Some(v.clone()),
+            None => None,
+        };
         // We are online, attempt the pw to the server.
-        match self.client.unix_user_authenticate(account_id, cred).await {
-            Ok(Some(mut n_tok)) => {
+        match self
+            .client
+            .unix_user_authenticate_step(account_id, ucred.as_deref())
+            .await
+        {
+            Ok(ProviderResult::UserToken(Some(mut n_tok))) => {
                 if self.check_nxset(&n_tok.name, n_tok.gidnumber).await {
                     // Refuse to release the token, it's in the denied set.
                     self.delete_cache_usertoken(n_tok.uuid).await?;
-                    Ok(None)
+                    Ok(ClientResponse::PamStatus(None))
                 } else {
                     debug!("online password check success.");
                     self.set_cache_usertoken(&mut n_tok).await?;
-                    self.set_cache_userpassword(n_tok.uuid, cred).await?;
-                    Ok(Some(true))
+                    match cred {
+                        Some(PamCred::Password(cred)) => {
```

Comment:
Here we've kept track of whether the id provider indicated that it was requesting a Password from the pam prompt. If it was a password, then we cache it for offline auth. If it was anything else (such as an MFA token), we ignore caching (otherwise we'll muck up the offline cache).